source/source_test.go

```diff
@@ -16,15 +16,474 @@ package source
 
 import (
 	"context"
+	"encoding/json"
 	"errors"
+	"fmt"
+	"os"
+	"strings"
 	"testing"
+	"time"
 
+	"cloud.google.com/go/storage"
 	"github.com/conduitio/conduit-connector-google-cloudstorage/config"
+	sourceConfig "github.com/conduitio/conduit-connector-google-cloudstorage/source/config"
+	"github.com/conduitio/conduit-connector-google-cloudstorage/source/position"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/google/uuid"
+	"google.golang.org/api/iterator"
+	"google.golang.org/api/option"
 )
 
+type Object struct {
+	key     string
+	content string
+}
+
+const (
+	projectID = "projectID"
+)
+
+func TestSource_SuccessfulSnapshot(t *testing.T) {
+	client, cfg := prepareIntegrationTest(t)
+
+	ctx := context.Background()
+	testBucket := cfg[config.ConfigKeyGCSBucket]
+	source := &Source{}
+	defer func() {
+		_ = source.Teardown(ctx)
+	}()
+
+	err := source.Configure(ctx, cfg)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = source.Open(ctx, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	testFiles := addObjectsToTheBucket(ctx, t, testBucket, client, 5)
+
+	// read and assert
+	for _, file := range testFiles {
+		_, err := readAndAssert(ctx, t, source, file)
+		if err != nil {
+			t.Fatalf("unexpected error: %v", err)
+		}
+	}
+
+	_, err = source.Read(ctx)
+	if !errors.Is(err, sdk.ErrBackoffRetry) {
+		t.Fatalf("expected a BackoffRetry error, got: %v", err)
+	}
+}
+
+func TestSource_SnapshotRestart(t *testing.T) {
+	client, cfg := prepareIntegrationTest(t)
+
+	ctx := context.Background()
+	testBucket := cfg[config.ConfigKeyGCSBucket]
+	source := &Source{}
+	defer func() {
+		_ = source.Teardown(ctx)
+	}()
+
+	err := source.Configure(ctx, cfg)
+	if err != nil {
+		t.Fatal(err)
+	}
+	p, err := json.Marshal(position.Position{
+		Key:       "file0003",
+		Timestamp: time.Now(),
+		Type:      0,
+	})
```

Comment:
Here, we are creating a position ourselves, which is not bad. However, to have a better integration test, we should use the position returned by the source. So, I'd suggest following:
1. put some objects into the bucket
2. start the source
3. expect the objects to be read
4. stop and tear down the source
5. put some more objects into the bucket
6. start the source
7. expect the new objects to be read

Reply:
In the above steps tear down happened in the snapshot mode so starting again will re-trigger the snapshot again resulting duplicates. Below steps would fit the scenario.
1.put some objects into the bucket
2.start the source
3.expect the objects to be read
4.add an object again to change it to CDC mode 
5.stop and tear down the source
6.put some more objects into the bucket
7.start the source **with the last position previously stored.**
8.expect the new objects to be read

Reply:
@WeirdMagician Exactly, I missed adding it to the list, even though I mentioned that we should use the position returned by the source.

Reply:
Completed Waiting for Approval to merge

Reply:
At least in this test function, I don't think I see the steps we discussed above. I still see a single Open() call and a hard-coded position.

Reply:
For the steps we have discussed i have created a another testcase now it is rename to TestSource_CDCRestart

Reply:
>  Here in this testcase the hard coded position is required to test the behavior of snapshot tests.

@WeirdMagician What is that test case actually? A restarted snapshot?

Reply:
Yes this behavior will happen when it got restarted from a non-nil snapshot position.
Here in this particular testcase we are not testing the entire connector restart use-case. Instead we are testing the behavior post restart called (SnapshotStartFromNonNilPosition).

Reply:
It would be good to have a full test case, i.e. one where you start a snapshot, stop it, then start again. But we can do that in a separate PR (if you agree with the idea in the first place).

Reply:
Add in the existing issue #23